In [3]:
import pandas as pd
import numpy as np


In [12]:
df = pd.read_csv(r"C:\Users\dell\Desktop\Data Analyst Project\zomato2\zomato.csv")

In [13]:
# Lets check the total number of restaurants in Bengaluru
df.shape

(51717, 17)

In [14]:
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Lets now check the total number of missing values in each column

In [16]:
# We can say that the URL of each restaurant in the dataset is our primary key
# So dropping all the data that are duplicate based on URL column

df = df.drop_duplicates('url')

In [17]:
df.shape

(51717, 17)

In [ ]:
 ## So no values were duplicate in our dataset

In [ ]:
# Cleaning the data column wise
# Phase I of cleaning
# Lets one by one check for each column, if there is any noise in the data and clean parallely if any


# URL Column

In [19]:
df['url'].isnull().sum()

0

In [ ]:
## No null values in our URL column, which was quite obvious since we scraped the data using the url itself

# Address Column

In [20]:
df['address'].isnull().sum()

0

In [ ]:
## Our Address column also does not have any missing values

# Name Column

In [21]:
df['name'].isnull().sum()

0

In [ ]:
##No null values in Name, don't you think this was also obvious, "A restaurant without a name". 
##Oh I think I got a name for my restaurant

# Online order column

In [24]:
df['online_order'].unique()

array(['Yes', 'No'], dtype=object)

In [ ]:
##No missing values in our Online_order column, contains only Yes or No

# Book Table Column

In [25]:
df['book_table'].unique()

array(['Yes', 'No'], dtype=object)

In [ ]:
## No missing values in our Book_table column, contains only Yes or No

# Rating Column

In [26]:
df['rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

In [ ]:
## So there are some missing values e.g. nan, -, lets encode them with some value or leave them blank

In [27]:
na_values = ['-', np.nan]

In [29]:
## Lets replace nan with blanks

df['rate'].replace(np.nan, '', regex=True, inplace=True)
df['rate'].replace('-', '', regex=True, inplace=True)
df['rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', '', '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

We have replaced all the nan and its equivalent by empty string or blanks which will now be our symbol for nans

In [31]:
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Votes Column

In [32]:
df['votes'].isnull().sum()

0

# Phone Column

In [34]:
# Lets check whether phone number contains any character other than numbers and '+'
df[~df['phone'].str.contains('[0-9+]',  na=False)]

url  \
121    https://www.zomato.com/bangalore/new-mangalore...   
234    https://www.zomato.com/bangalore/kadamba-class...   
300    https://www.zomato.com/bangalore/just-thindi-b...   
378    https://www.zomato.com/bangalore/dakshin-kitch...   
466    https://www.zomato.com/bangalore/chettys-corne...   
...                                                  ...   
51353  https://www.zomato.com/bangalore/chatpata-indi...   
51501  https://www.zomato.com/bangalore/toastys-white...   
51591  https://www.zomato.com/bangalore/donne-biryani...   
51641  https://www.zomato.com/bangalore/andhra-spices...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   

                                                 address  \
121    89, Near PES College, 22nd Main Road, Girinaga...   
234    Satellite Bus Stand, KSRTC, Mysore Road, Banga...   
300    27, Gandhi Bazaar Main Rd, Basavanagudi, Banga...   
378    Kabirmutt Road, Gavipuram Extension, Basavanag...   
466    126/3, Uttarahalli Main Road, Sarvabouma Nagar...   
...                                                  ...   
51353  Food Court, Ascendas Park Square Mall, ITPL Ma...   
51501  3rd Floor, Virginia Mall, Varthur Main Road, W...   
51591  26, Aparaajita Building, Borewell Road, Opposi...   
51641  188, ITPL Main Road, Brookefields, Bangalore, ...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   

                                                    name online_order  \
121                             New Mangalore Lunch Home           No   
234                                      Kadamba Classic           No   
300                                          Just Thindi          Yes   
378                                      Dakshin Kitchen          Yes   
466                                      Chetty's Corner          Yes   
...                                                  ...          ...   
51353                                     Chatpata India          Yes   
51501                                           Toasty's           No   
51591                                 Donne Biryani Mane          Yes   
51641                                      Andhra Spices           No   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   

      book_table    rate  votes phone      location    rest_type  \
121           No   3.3/5      7   NaN  Banashankari  Quick Bites   
234           No   3.6/5     16   NaN   Mysore Road  Quick Bites   
300           No   3.6/5     16   NaN  Basavanagudi  Quick Bites   
378           No   3.6/5     15   NaN  Basavanagudi  Quick Bites   
466           No   3.4/5     30   NaN   Uttarahalli  Quick Bites   
...          ...     ...    ...   ...           ...          ...   
51353         No  3.9 /5    103   NaN    Whitefield   Food Court   
51501         No              0   NaN    Whitefield   Food Court   
51591         No  3.5 /5     51   NaN    Whitefield  Quick Bites   
51641         No  3.2 /5      4   NaN   Brookefield  Quick Bites   
51714         No              0   NaN    Whitefield          Bar   

               dish_liked                                 cuisines  \
121                   NaN  Mangalorean, South Indian, North Indian   
234                   NaN                             South Indian   
300                   NaN                             South Indian   
378                   NaN                             South Indian   
466                   NaN                 Fast Food, Burger, Pizza   
...                   ...                                      ...   
51353  Chaat, Raj Kachori                             North Indian   
51501                 NaN                                Fast Food   
51591                 NaN                                  Biryani   
51641                 NaN            Andhra, Chinese, South Indian   
51714                 NaN                              Finger Food   

      approx_cost(for two people)  \
121                           200   
23

So there are 1,208 data that does not have a phone number. Lets replace them by blank.

In [35]:
df[~df['phone'].str.contains('[0-9+]',  na=False)]['phone'].unique()

array([nan], dtype=object)

In [37]:
df['phone'].replace(np.nan, '', regex=True, inplace=True)

In [38]:
df[~df['phone'].str.contains('[0-9+]',  na=False)]['phone'].unique()

array([''], dtype=object)

All the Phone numbers are replaced by blank...

# Location Column

In [39]:
df['location'].isnull().sum()

21

In [ ]:
## There are 21 nan values in location lets replace them with blank

In [40]:
df['location'].replace(np.nan, '', regex=True, inplace=True)

In [41]:
df['location'].isnull().sum()

0

Perfect !!! There no Nan anymore

# Rest type column

In [43]:
df['rest_type'].isnull().sum()

227

In [44]:
## Lets replace with blank

df['rest_type'].replace(np.nan, '', regex = True,inplace=True)

In [45]:
df['rest_type'].isnull().sum()

0

# Dished liked

In [46]:
df['dish_liked'].isnull().sum()

28078

In [47]:
##There is huge data of null in dish liked

df['dish_liked'].replace(np.nan, '', regex=True, inplace=True)
df['dish_liked'].isnull().sum()


0

# Cuisines column

In [48]:
df['cuisines'].isnull().sum()

45

In [49]:
## replace with blank
df['cuisines'].replace(np.nan, '', regex=True, inplace=True)
df['cuisines'].isnull().sum()


0

# Approx_cost column

In [52]:
# Lets check the number of null values in the column
print("NULL values : ",df["approx_cost(for two people)"].isnull().sum())

NULL values :  346


In [54]:
df['approx_cost(for two people)'].replace(np.nan, '', regex=True, inplace=True)
df['approx_cost(for two people)'].isnull().sum()


0

# Reviews_list column

In [55]:
# lets check the NULL values in Reviews_list column
print("NULL values : ",df['reviews_list'].isnull().sum())
print("Blank : ", (df['reviews_list']=="[]").sum())

NULL values :  0
Blank :  7595


7595 Review_list are empty

In [56]:
df[df['reviews_list']=="[]"].head()

url  \
75   https://www.zomato.com/bangalore/om-sri-vinaya...   
91   https://www.zomato.com/bangalore/sunsadm-banas...   
92   https://www.zomato.com/bangalore/annapooranesh...   
95   https://www.zomato.com/bangalore/vi-ras-bar-an...   
107  https://www.zomato.com/bangalore/coffee-shopee...   

                                               address  \
75   39, S B M Colony, Near-Sita Circle, 1st Stage,...   
91   Shop 18 , 3rd cross road,Markandeya Layout, Ba...   
92   625, 5th Cross Road, Hanumanthnagar, Banashank...   
95   3828, Near Deve Gowda Petrol Bunk, 30th Cross,...   
107  1722, 13th Main Road 31st Cross BSK, 2nd Stage...   

                           name online_order book_table   rate  votes  \
75        Om Sri Vinayaka Chats           No         No    NEW      0   
91                      Sunsadm           No         No             0   
92       Annapooraneshwari Mess           No         No             0   
95   Vi Ra's Bar and Restaurant           No         No  3.3/5     62   
107               Coffee Shopee           No         No             0   

                                phone      location           rest_type  \
75                     +91 8553206035  Banashankari  Takeaway, Delivery   
91                     +91 7829290358  Banashankari  Takeaway, Delivery   
92                     +91 9742009518  Banashankari                Mess   
95   +91 9845063930\r\n+91 7411148111  Banashankari                 Bar   
107                    +91 9591504624  Banashankari  Takeaway, Delivery   

         dish_liked                             cuisines  \
75                                Street Food, Fast Food   
91                                          South Indian   
92                                          South Indian   
95   Chilli Chicken  South Indian, North Indian, Chinese   
107                                            Beverages   

    approx_cost(for two people) reviews_list menu_item listed_in(type)  \
75                          500           []        []        Delivery   
91                          400           []        []        Delivery   
92                          200           []        []        Delivery   
95                          800           []        []        Delivery   
107                         250           []        []        Delivery   

    listed_in(city)  
75     Banashankari  
91     Banashankari  
92     Banashankari  
95     Banashankari  
107    Banashankari

Reviews_list is the list of tuples. Each tuple consists of two values. One is the rating and the other is the review. Since after reading from the file they would have been converted back to string. We need to convert it back to list of tuple format

In [58]:
import ast
df['reviews_list'] =  df['reviews_list'].apply(ast.literal_eval)
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [ ]:
## Lets count the number of reviews in our dataset

In [59]:
df['reviews_list'].apply(len).sum()

1320051

We have 1,320,051 reviews in our datset

# Menu Item

In [60]:
# lets check the NULL values in Menu_item column
print("NULL values : ",df['menu_item'].isnull().sum())
print("Blank : ", (df['menu_item']=="[]").sum())

NULL values :  0
Blank :  39617


There are 39,617 Menu_item that are empty. The menu item is a list of menus which are right now in string format. Lets convert them back to list format

In [61]:
df['menu_item'] = df['menu_item'].apply(ast.literal_eval)
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Listed_in(type) column

In [64]:
# lets check the NULL values in Listed_in(type) column
print("NULL values : ",df['listed_in(type)'].isnull().sum())
print("Blank : ", (df['listed_in(type)']=="").sum())

NULL values :  0
Blank :  0


# Listed_in(city) column

In [65]:
# lets check the NULL values in Listed_in(type) column
print("NULL values : ",df['listed_in(city)'].isnull().sum())
print("Blank : ", (df['listed_in(city)']=="").sum())

NULL values :  0
Blank :  0


# Phase II of cleaning
On going through the data once we found no noise or anamoly in our URL, Address, Name, Online_order, Book_table, Rate, Votes column.

In [ ]:
##Phone column
##In the Phone column we saw that if the restaurant has multiple phone number they are separated by \r\n or \n. 
## Lets replace all those phone number that contains letters other than space character, 0-9, \r, \n, plus(+) by blank.

In [66]:
df['phone'] = df['phone'].str.replace('[^\+0-9\r\n\s]','')

C:\Users\dell\AppData\Local\Temp\ipykernel_11372\3139335625.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['phone'] = df['phone'].str.replace('[^\+0-9\r\n\s]','')


Lets check if any column contains phone other than our permissible character.

In [67]:
df['phone'].str.contains('[^\+0-9\r\n\s]',na = False).sum()

0

None Perfect !!!

# Rest_type column 

In [68]:
df['rest_type'].unique()

array(['Casual Dining', 'Cafe, Casual Dining', 'Quick Bites',
       'Casual Dining, Cafe', 'Cafe', 'Quick Bites, Cafe',
       'Cafe, Quick Bites', 'Delivery', 'Mess', 'Dessert Parlor',
       'Bakery, Dessert Parlor', 'Pub', 'Bakery', 'Takeaway, Delivery',
       'Fine Dining', 'Beverage Shop', 'Sweet Shop', 'Bar',
       'Beverage Shop, Quick Bites', 'Confectionery',
       'Quick Bites, Beverage Shop', 'Dessert Parlor, Sweet Shop',
       'Bakery, Quick Bites', 'Sweet Shop, Quick Bites', 'Kiosk',
       'Food Truck', 'Quick Bites, Dessert Parlor',
       'Beverage Shop, Dessert Parlor', 'Takeaway', 'Pub, Casual Dining',
       'Casual Dining, Bar', 'Dessert Parlor, Beverage Shop',
       'Quick Bites, Bakery', 'Dessert Parlor, Quick Bites',
       'Microbrewery, Casual Dining', 'Lounge', 'Bar, Casual Dining',
       'Food Court', 'Cafe, Bakery', '', 'Dhaba',
       'Quick Bites, Sweet Shop', 'Microbrewery',
       'Food Court, Quick Bites', 'Pub, Bar', 'Casual Dining, Pub',
       

It looks clean already !

In [69]:
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Location column

In [70]:
df['location'].unique()

array(['Banashankari', 'Basavanagudi', 'Mysore Road', 'Jayanagar',
       'Kumaraswamy Layout', 'Rajarajeshwari Nagar', 'Vijay Nagar',
       'Uttarahalli', 'JP Nagar', 'South Bangalore', 'City Market',
       'Nagarbhavi', 'Bannerghatta Road', 'BTM', 'Kanakapura Road',
       'Bommanahalli', '', 'CV Raman Nagar', 'Electronic City', 'HSR',
       'Marathahalli', 'Sarjapur Road', 'Wilson Garden', 'Shanti Nagar',
       'Koramangala 5th Block', 'Koramangala 8th Block', 'Richmond Road',
       'Koramangala 7th Block', 'Jalahalli', 'Koramangala 4th Block',
       'Bellandur', 'Whitefield', 'East Bangalore', 'Old Airport Road',
       'Indiranagar', 'Koramangala 1st Block', 'Frazer Town', 'RT Nagar',
       'MG Road', 'Brigade Road', 'Lavelle Road', 'Church Street',
       'Ulsoor', 'Residency Road', 'Shivajinagar', 'Infantry Road',
       'St. Marks Road', 'Cunningham Road', 'Race Course Road',
       'Commercial Street', 'Vasanth Nagar', 'HBR Layout', 'Domlur',
       'Ejipura', 'Jeevan B

In [71]:
df['location'].nunique()

94

So there are 93 neighborhoods in Bengaluru, 94rd is nan

# Listed_in(type) column

In [72]:
df['listed_in(type)'].unique()

array(['Buffet', 'Cafes', 'Delivery', 'Desserts', 'Dine-out',
       'Drinks & nightlife', 'Pubs and bars'], dtype=object)

Already Sorted

# Listed_in(city) column

In [74]:
df['listed_in(city)'].unique()

array(['Banashankari', 'Bannerghatta Road', 'Basavanagudi', 'Bellandur',
       'Brigade Road', 'Brookefield', 'BTM', 'Church Street',
       'Electronic City', 'Frazer Town', 'HSR', 'Indiranagar',
       'Jayanagar', 'JP Nagar', 'Kalyan Nagar', 'Kammanahalli',
       'Koramangala 4th Block', 'Koramangala 5th Block',
       'Koramangala 6th Block', 'Koramangala 7th Block', 'Lavelle Road',
       'Malleshwaram', 'Marathahalli', 'MG Road', 'New BEL Road',
       'Old Airport Road', 'Rajajinagar', 'Residency Road',
       'Sarjapur Road', 'Whitefield'], dtype=object)

In [75]:
df['listed_in(city)'].nunique()

30

Listed_in(city) contains the restaurants data listed by city in the zomato bangalore main page. Although there are'nt only above mentioned neighborhoods in bangalore. But the data in the website was listed by cities this way. There were 30 neighborhoods which were separately listed.

In [77]:
df.head()


url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Phase III of cleaning
Assigning appropriate data type and column name to our data

In [78]:
# Verifying data types for each column
df.dtypes

url                            object
address                        object
name                           object
online_order                   object
book_table                     object
rate                           object
votes                           int64
phone                          object
location                       object
rest_type                      object
dish_liked                     object
cuisines                       object
approx_cost(for two people)    object
reviews_list                   object
menu_item                      object
listed_in(type)                object
listed_in(city)                object
dtype: object

For each of the column assignng their appropriate categories based on the analysis we have done so far.

In [79]:
df['online_order'] = df['online_order'].astype('category')
df['book_table'] = df['book_table'].astype('category')
df['location'] = df['location'].astype('category')
df['rest_type'] = df['rest_type'].astype('category')
df['listed_in(type)'] = df['listed_in(type)'].astype('category')
df['listed_in(city)'] = df['listed_in(city)'].astype('category')

In [80]:
df.dtypes

url                              object
address                          object
name                             object
online_order                   category
book_table                     category
rate                             object
votes                             int64
phone                            object
location                       category
rest_type                      category
dish_liked                       object
cuisines                         object
approx_cost(for two people)      object
reviews_list                     object
menu_item                        object
listed_in(type)                category
listed_in(city)                category
dtype: object

One thing to be worth noting is the same restaurant can be listed in different category with different URL. hence URL cannot be the primary key

# That's all about cleaning of  this dataset